### PR TITLE
Android support for recording video

### PR DESCRIPTION
**Pull request by marcejohnson**
Android support for recording video:

- Same parameters and values as IOS, including captureMode, captureTarget, captureQuality, etc.
- Updated Readme.md
- Example code

"Drop in" for compatibility with IOS... to make Android and IOS equivalent.  I will continue to enhance this as bugs or additional changes are needed since I need this for my own project.  Please send me you feedback, thanks.

To install latest version, run:
```
npm install --save git+https://github.com/marcejohnson/react-native-camera.git#master
```

### CODE REVIEW

**Comment by vespakoen** [comment]
Awesome!!

Thank you so much for putting that code online, I have been banging my head against the wall to make this work.

I think we can combine both forks and get something good out of it!

**Comment by marcejohnson** [comment]
Thanks, i hope this unblocks you and others.  This is so exciting to finally be approaching parity with iOS on video recording.    I'll review your code and share my thoughts.   Feel free to do the same.  Opportunities for improvement in mine include camera roll, camera flip, and audio only stream.  If you have code that handles that better, that would help.. I am also interested in if I chose the best way to structure the code.

File: android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java [hunk 1]
```diff
@@ -162,6 +170,157 @@ public String getName() {
...
+                videoFile = getOutputMediaFile(MEDIA_TYPE_VIDEO);
+                break;
+        }
+        if (videoFile == null) {
+            return false;
+        }
+        mediaRecorder.setOutputFile(videoFile.getPath());
+
+        mediaRecorder.setMaxDuration(600000); // Set max duration 60 sec.
+        mediaRecorder.setMaxFileSize(50000000); // Set max file size 50M
```
Comment (dhrrgn) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java]: These should be settable in the `options` map sent to `record`.

Comment (marcejohnson) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java]: @dhrrgn :  done

**Comment by marcejohnson** [comment]
Got in all the improvements and features requested.  Runs well in my test apps.  Still no conflicts on base branch.  Let me know what you think.

**Comment by vespakoen** [comment]
Great! Looks like my changes are irrelevant now (they all seem to be in this pull)

I had an orientation bug using your pull request (long time ago, probably fixed?) the recorded video was upside down.

Another thing I noticed when working on this is that when the maxDuration is less than 4 seconds, it doesn't work (threw an exception I think? cannot remember exactly...)

I would love to be able to record a 1 second clip and I think we might be able to work around it by using our own timer and stop the recording after X seconds. (perhaps only when the duration < 4 secs?)

Will test this out in a bit.

**Comment by marcejohnson** [comment]
HI Vespakoen, thanks for reviewing!  Yes please pull the latest code (as you did) -- there is code in there to handle orientation on different platforms.  Regarding the minimum recording limitation on Android:  in my testing I found that less than 1.5 seconds threw an exception, so that might be the minimum.  But if its less < 4 on your system, can add that minimum in there trivially.

**Comment by vespakoen** [comment]
Great!

Will test it later, unfortunately I am busy with tax administration now, and be working after that, so I can only review tomorrow (in about 20 hours).

File: android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java [hunk 2]
```diff
@@ -54,11 +59,27 @@
 
     private final ReactApplicationContext _reactContext;
 
+    private MediaRecorder mediaRecorder = new MediaRecorder();
```
Comment (vespakoen) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java]: Non-public, non-static field names start with m.

http://stackoverflow.com/questions/2092098/why-do-most-variables-in-android-tutorial-start-with-m

Not that it matters much ;)

File: android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java [hunk 3]
```diff
@@ -162,6 +183,191 @@ public String getName() {
...
+        if (videoFile == null) {
+            return false;
+        }
+
+        mediaRecorder.setOutputFile(videoFile.getPath());
+
+        // Attach callback for maxDuration and maxFileSize
+        mediaRecorder.setOnInfoListener(this);
+ 
+        if (options.hasKey("maxDuration")) {
```
Comment (vespakoen) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java]: `maxDuration` should be `totalSeconds`, that's what it is called on IOS at least...

File: android/src/main/java/com/lwansbrough/RCTCamera/RCTCamera.java [hunk 4]
```diff
@@ -109,6 +114,36 @@ public int getPreviewHeight(int type) {
...
+    {
+        Camera camera = _cameras.get(type);
+        Camera.Size result = null;
+        if(camera == null) {
+            return null;
+        }
+        Camera.Parameters params = camera.getParameters();
+        // defer to preview instead of params.getSupportedVideoSizes() http://bit.ly/1rxOsq0
+        List<Size> sizes = params.getSupportedPreviewSizes();
+
```
Comment (vespakoen) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCamera.java]: `params.getSupportedVideoSizes();` should be preferred! (pull request coming up ;))

**Comment by vespakoen** [comment]
Quick summary of the latest changes:

This pull now also works on my Galaxy S4 (no more: `start failed -19`), although the orientation is still flipped.

marcejohnson reported that it works well on all his devices, I will test it on my Nexus tablet tomorrow.

I was also able to record a short 1 second video with it.

Basically I am really happy with this pull request, it has come a long way already and I think it's ready to merge.

**Comment by marcejohnson** [comment]
I agree with @vespakoen.  This gets us from zero to one on Android video capture after lots of work and review.  Unless anyone has any changes, recommend merging this in so we can get this supported.  We'll be responsive to change requests.  @lwansbrough

**Comment by vespakoen** [comment]
Another thing I noticed `RCTCamera`

- There is some duplication in the `get*PictureSize` and `get*VideoSize` methods.
The only difference I see is the `if (null != sizes) {` check, which is probably not needed anymore now that we use the `getSupportedSizesWithFallback` to get the sizes.

- The "medium" setting still uses `getSupportedPreviewSizes`, should probably use `getSupportedSizesWithFallback` and compare that with `getSmallestVideoSize` and NOT devide the sizes by 2 in case they are the same.

- Orientation / rotation is all over the place, I feel like this could be greatly simplified
Perhaps the fine people from this pull request can shine more light on the matter: https://github.com/lwansbrough/react-native-camera/pull/271

**Comment by vespakoen** [comment]
@markomatic can you shine some light on this?
It looks like you made a lot of changes to the preview / orientation / rotation handling.
What is the reason for not updating them in this repository?

Your implementation looks a lot cleaner to me.

**Comment by marcejohnson** [comment]
Also I notice in the @latest build  "capture now returns an object instead of a string".  So we might want to update to support that if making more changes.

**Comment by vespakoen** [comment]
I found some good examples on the internet using Camera2.

I am going to work on implementing that over here:

https://github.com/vespakoen/react-native-camera2

This is a longer-term / hobby project, so don't expect anything useful to come out of it soon.

It is based on a google sample from here: https://github.com/googlesamples/android-Camera2Raw/tree/master/Application/src/main/java/com/example/android/camera2raw

I got the preview working, now I am looking for a nice way to communicate from the `view` -> `module` and the other way around.

I have tried passing one to the other from the `package`, that didn't work, so I guess I have to resort to a thing similar to `RCTCamera.getInstance`, ahh, anyways, I will keep you guys up to date on the progress, but expect it to take a long time ;) (unless you help out of course, haha)

**Comment by afterbugs** [comment]
You bring up a good point.  I'm not sure why this project chose Camera instead of Camera2 at first.  I think people were holding off (in the past) because older devices didn't support Camera2, but now it has been around for years and has good device compatibility-- and in some cases better capabilities on newer devices.  If switching to Camera2 helps, I will totally help.  Looks like you pulled the existing code over.  Please take an initial run at it (like you are) and I'll hop in when you get stuck.  Maybe together we can get this done sooner rather than later.

**Comment by marcejohnson** [comment]
(reposting using correct account)
You bring up a good point. I'm not sure why this project chose Camera instead of Camera2 at first. I think people were holding off (in the past) because older devices didn't support Camera2, but now it has been around for years and has good device compatibility-- and in some cases better capabilities on newer devices. If switching to Camera2 helps, I will totally help. Looks like you pulled the existing code over. Please take an initial run at it (like you are) and I'll hop in when you get stuck. Maybe together we can get this done sooner rather than later.

**Comment by vespakoen** [comment]
Yeah that could very well be the reason.

At the moment Camera is deprecated, so I guess it makes sense to move over to Camera2.

Another advantage is that there is example code (from google) which is very likely to contains less bugs and work more reliably.

There is code for picture handling in there, code for video recording can be "stolen" from here: https://github.com/googlesamples/android-Camera2Video/blob/master/Application/src/main/java/com/example/android/camera2video/Camera2VideoFragment.java

Then it's a matter of making it compile and glue it to react-native.

I am also trying to get rid of the `aspect fit` and `orientation` properties on the camera preview, (which should keep the code simpler) over here: https://github.com/lwansbrough/react-native-camera/issues/274

File: android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java [hunk 5]
```diff
@@ -162,6 +183,185 @@ public String getName() {
...
+        mMediaRecorder.setAudioSource(MediaRecorder.AudioSource.CAMCORDER);
+        mMediaRecorder.setVideoSource(MediaRecorder.VideoSource.CAMERA);
+
+        int mediaRecorderHintOrientation = (90 + ((720 - RCTCamera.getInstance().getActualDeviceOrientation() * 90))) % 360;
+
+        // adjust for differences in how devices display front facing http://www.theverge.com/2015/11/9/9696774/google-nexus-5x-upside-down-camera
+        if (RCT_CAMERA_TYPE_FRONT == options.getInt("type")) {
+            if ((RCT_CAMERA_ORIENTATION_PORTRAIT == RCTCamera.getInstance().getOrientation()) && (RCTCamera.getInstance().getAdjustedDeviceOrientation() == 90)) {
+                mediaRecorderHintOrientation += 180;
+            }
```
Comment (Phoenix1708) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java]: I see the 2 way front camera mounting issue, but I don't understand why only "trying to fix the upside down issue" when in portrait ? I tested on Nexus 5. The front camera recording is upside down but if I go landscape, the .mp4 recording is in the correct orientation.

Great work BTW. Thank you for adding this !

**Comment by bradhanebury** [comment]
Since I know you guys are still working on this, I've been fiddling with it on a current project. Have run into the following issue when trying to stop recording a video. 

```
E/unknown:React: Exception in native call from JS
                                                            java.lang.RuntimeException: getParameters failed (empty parameters)
                                                                at android.hardware.Camera.native_getParameters(Native Method)
                                                                at android.hardware.Camera.getParameters(Camera.java:1953)
                                                                at com.lwansbrough.RCTCamera.RCTCamera.setCaptureVideoQuality(RCTCamera.java:273)
                                                                at com.lwansbrough.RCTCamera.RCTCameraModule.prepareMediaRecorder(RCTCameraModule.java:187)
                                                                at com.lwansbrough.RCTCamera.RCTCameraModule.record(RCTCameraModule.java:269)
                                                                at com.lwansbrough.RCTCamera.RCTCameraModule.capture(RCTCameraModule.java:375)
                                                                at java.lang.reflect.Method.invoke(Native Method)
                                                                at java.lang.reflect.Method.invoke(Method.java:372)
                                                                at com.facebook.react.bridge.BaseJavaModule$JavaMethod.invoke(BaseJavaModule.java:268)
                                                                at com.facebook.react.bridge.NativeModuleRegistry$ModuleDefinition.call(NativeModuleRegistry.java:185)
                                                                at com.facebook.react.bridge.NativeModuleRegistry.call(NativeModuleRegistry.java:64)
                                                                at com.facebook.react.bridge.CatalystInstanceImpl$NativeModulesReactCallback.call(CatalystInstanceImpl.java:420)
                                                                at com.facebook.react.bridge.queue.NativeRunnableDeprecated.run(Native Method)
                                                                at android.os.Handler.handleCallback(Handler.java:739)
                                                                at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:31)
                                                                at android.os.Looper.loop(Looper.java:145)
                                                                at com.facebook.react.bridge.queue.MessageQueueThreadImpl$3.run(MessageQueueThreadImpl.java:187)
                                                                at java.lang.Thread.run(Thread.java:818)
```

**Comment by bradhanebury** [comment]
Also ran into an issue in RCTCameraViewFinder.java's startCamera method. The getBestPictureSize method was not returning a valid size for my phone type. I ran a test and set the parameters manually via `parameters.setPictureSize(1280, 720);` and it worked fine. 

Device is a Samsung Galaxy S4 running Android 5.0.1

**Comment by Phoenix1708** [comment]
@marcejohnson Hi, I updated my comments on the media recorder orientation. Could you please check it out ? Thanks !

**Comment by tannerkrewson** [comment]
Hello everyone,

If anyone else was trying to get marcejohnson's fork to work without success, I have made a pull request the fixes the issues by adding the essential changes that were made to the main repo:

https://github.com/marcejohnson/react-native-camera/pull/2

I would love to go through and resolve the conflicts to get this merged into the main repo, however I don't have a Mac to test iOS on.

**Comment by marcejohnson** [comment]
@tannerkrewson thanks.  I merged your pull request.

**Comment by marcejohnson** [comment]
@Phoenix1708 thanks for your comment.  Do you have a code change for this?  @tannerkrewson you said "I would love to go through and resolve the conflicts to get this merged into the main repo, however I don't have a Mac to test iOS on":  I have a mac/IOS to regression test on, so can help here if you want to take a crack at the merge.  Cheers.

**Comment by faceyspacey** [comment]
What's the status of this? What's holding it back from being merged?

**Comment by marcejohnson** [comment]
Since master has moved on, this branch has merge conflicts that must be resolved (hopefully not a big deal).  Folks above have offered to resolve the conflicts to get this merged into the main repo.  @faceyspacey if you've done this, please submit your pull request and I'll review/test.  Otherwise I could take a run at it if no one else is doing it.

**Comment by sscaff1** [comment]
Any updates on this merge?

**Comment by jacob-israel-turner** [comment]
Any updates here?  I need this for a small project I'm working on now, which will need to be completed in a couple of weeks.  I can help review the PR if I have a little bit of direction.

**Comment by markomatic** [comment]
@charpeni, conflicts are resolved in PR https://github.com/marcejohnson/react-native-camera/pull/3. After it is merged, this PR can be merged without conflicts.

**Comment by marcejohnson** [comment]
Thanks @markomatic really appreciated!
I did a quick review of the diff (https://github.com/marcejohnson/react-native-camera/pull/3/files).  I notice extra files in the merge unrelated to this android component.   E.g.  ios/*, app/BUCK, example/* and miscellaneous development environment files.  Any need to include these?

Anyone who needs this - please do a quick test and sign off on the PR marcejohnson#3 and report back here.   The more people who test, the better chance of getting merged to main repo.  I am on a car trip through tomorrow and I will also try to test tomorrow when I have a stable internet connection.  As soon as tested (and possibly with the extra files pulled out) will do the merge.

**Comment by markomatic** [comment]
@marcejohnson, thanks. All changes in ios/, app/BUCK and Example/ are from latest lwansbrough/react-native-camera master, as this PR is going to master. I tested android and iOS build (It would be excelent for others to test it also). Example is nothing but sample app using camera module (I added camera recording there, so I can test it before PR). Whoever added BUCK files, please test them in PR https://github.com/marcejohnson/react-native-camera/pull/3, but it should work as it is only part of merge commit.

**Comment by marcejohnson** [comment]
Thanks @markomatic. I pulled the changes into my Android project, and it worked without regression.  Merged.  @charpeni, conflicts are resolved and this well vetted PR can be merged without conflicts... to finally add "Android support for recording video"!

File: android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java [hunk 6]
```diff
@@ -293,44 +510,69 @@ public void hasFlash(ReadableMap options, final Promise promise) {
...
             String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss").format(new Date());
-            File outputDir = _reactContext.getCacheDir();
+            File outputDir;
             File outputFile;
 
             if (type == MEDIA_TYPE_IMAGE) {
+                outputDir = _reactContext.getCacheDir(); // for backwards compatibility
                 outputFile = File.createTempFile("IMG_" + timeStamp, ".jpg", outputDir);
             } else if (type == MEDIA_TYPE_VIDEO) {
+                outputDir = _reactContext.getFilesDir(); // for compatibility with react-native-fs
```
Comment (corbt) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java]: Why not use the tmp dir here? If I'm saving videos with the "temp" option, I don't expect to have to clean them up later, but this line requires me to.

Comment (markomatic) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java]: I agree. @marcejohnson, could you help us with "compatibility with react-native-fs"? These video files will not be cleaned up, as they will be saved into apps private files dir, not temporary dir?

File: android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java [hunk 7]
```diff
@@ -175,6 +194,185 @@ public String getName() {
...
+        File f = new File(mVideoFile.getPath());
+        if (f.exists()) {
+            f.setReadable(true, false); // so mediaplayer can play it
+            f.setWritable(true, false); // so can clean it up
+
+            if (mRecordingPromise != null) {
+                switch (mRecordingOptions.getInt("target")) {
+                    case RCT_CAMERA_CAPTURE_TARGET_MEMORY:
+                        byte[] encoded = convertFileToByteArray(mVideoFile);
+                        mRecordingPromise.resolve(new String(encoded, Base64.DEFAULT));
```
Comment (corbt) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java]: This should be `resolve`d with the new capture return options.

https://github.com/lwansbrough/react-native-camera#captureoptions-promise

**Comment by corbt** [comment]
@marcejohnson can you confirm that videos recorded on Android with your fork work? I can't seem to get the example project to record videos successfully -- they generate video files (which show up in Google Photos on the camera roll) but the files are very small, and the photos app says they're 0 seconds long and can't be played. Any insight?

**Comment by marcejohnson** [comment]
@corbt I can confirm I was able to record and playback using the "temp" target in my app.  Did that work for you?

**Comment by corbt** [comment]
@marcejohnson No, the `temp` target doesn't work, nor does disk. :( I'm using a Nexus 9, and have had video recording in iOS working for a long time.

Does the example in this repository work for you with recording video on Android?

**Comment by marcejohnson** [comment]
I think @markomatic added the "example" code recently.  @markomatic, if so,  can you please shed light on this for @corbt?  Thanks.

**Comment by corbt** [comment]
@marcejohnson would you mind copying in here your working `<Camera />` elements and call to `camera.capture()`? Perhaps I need to change something in my configuration of those elements.

**Comment by markomatic** [comment]
Yes, I added video recording sample into example inside Example directory. It uses camera roll as target (you can change target on whatever you want in https://github.com/lwansbrough/react-native-camera/pull/262/files#diff-8e927151f271a81709e2e70c2953591bR66). I tested this example on Xperia T (Android 4.1) and Xperia Z Ultra (Android 5.1). It worked on both devices.

If you want to run Example, then run from repository root through cli:
1. `cd Example/`
2. `npm install`
3. `adb reverse tcp:8081 tcp:8081` or in Dev Settings input COMPUTER_IP:8081 for debug server 
4. `react-native run-android`

It should work, if it's not working then it could be problem with Android 6.0, if you are running Android 6.0 on Nexus 9. I don't have any device with Android 6.0, so I can't test it there. If it's problem on Android 6.0 then we should wait for someone with that OS for testing.

File: example.index.android.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
+'use strict';
```
Comment (markomatic) [inline example.index.android.js]: This file can be removed as we have Example directory with camera app sample.

**Comment by tannerkrewson** [comment]
I tested the @markomatic's example on my Nexus 6 running 6.0. The video recorded correctly and was not 0 seconds long like @corbt's. The video is really choppy, but I don't think that is related.

I will gladly test anything else if anyone wants me to.

**Comment by corbt** [comment]
Yes I'm able to run the example successfully and take pictures, but taking a video just generates a very small file (16-300kb) that can't be played. I'll continue investigating, and see if it's a problem with my device or the installation. I'm testing on @marcejohnson's `master` branch.

**Comment by corbt** [comment]
Very strange, I've tested @markomatic's example for video recording on both a Nexus 9 and Oneplus One (both Android 6.0.1). Both demonstrate the same behavior I mentioned above -- pictures work fine, but for video a very small file is created on disk that can't actually be played.

**Comment by markomatic** [comment]
@corbt, could you open android studio and see if there is any error in logcat when you record video?

**Comment by corbt** [comment]
@markomatic Yes, here is the adb logcat output after recording/stopping a video in the example:

```
08-10 10:42:28.674 4441-4758/com.example E/MediaRecorder: stop failed: -1007
08-10 10:42:28.676 4441-4758/com.example W/System.err: java.lang.RuntimeException: stop failed.
08-10 10:42:28.681 4441-4758/com.example W/System.err:     at android.media.MediaRecorder.stop(Native Method)
08-10 10:42:28.681 4441-4758/com.example W/System.err:     at com.lwansbrough.RCTCamera.RCTCameraModule.releaseMediaRecorder(RCTCameraModule.java:307)
08-10 10:42:28.681 4441-4758/com.example W/System.err:     at com.lwansbrough.RCTCamera.RCTCameraModule.stopCapture(RCTCameraModule.java:495)
08-10 10:42:28.681 4441-4758/com.example W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
08-10 10:42:28.681 4441-4758/com.example W/System.err:     at com.facebook.react.bridge.BaseJavaModule$JavaMethod.invoke(BaseJavaModule.java:316)
08-10 10:42:28.681 4441-4758/com.example W/System.err:     at com.facebook.react.bridge.NativeModuleRegistry$ModuleDefinition.call(NativeModuleRegistry.java:187)
08-10 10:42:28.682 4441-4758/com.example W/System.err:     at com.facebook.react.bridge.NativeModuleRegistry.call(NativeModuleRegistry.java:62)
08-10 10:42:28.683 4441-4758/com.example W/System.err:     at com.facebook.react.bridge.CatalystInstanceImpl$NativeModulesReactCallback.call(CatalystInstanceImpl.java:432)
08-10 10:42:28.683 4441-4758/com.example W/System.err:     at com.facebook.react.bridge.queue.NativeRunnableDeprecated.run(Native Method)
08-10 10:42:28.683 4441-4758/com.example W/System.err:     at android.os.Handler.handleCallback(Handler.java:739)
08-10 10:42:28.683 4441-4758/com.example W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:95)
08-10 10:42:28.683 4441-4758/com.example W/System.err:     at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:31)
08-10 10:42:28.683 4441-4758/com.example W/System.err:     at android.os.Looper.loop(Looper.java:148)
08-10 10:42:28.683 4441-4758/com.example W/System.err:     at com.facebook.react.bridge.queue.MessageQueueThreadImpl$3.run(MessageQueueThreadImpl.java:187)
08-10 10:42:28.683 4441-4758/com.example W/System.err:     at java.lang.Thread.run(Thread.java:818)
08-10 10:42:28.774 4441-4757/com.example I/ReactNativeJS: content://media/external/video/media/3100
```

I'm investigating now, but any pointers would be helpful!

**Comment by markomatic** [comment]
Looks like there are results: https://www.google.com/search?q=android+camera+stop+failed+1007

File: android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java [hunk 9]
```diff
@@ -175,6 +194,185 @@ public String getName() {
...
+    @ReactMethod
+    private void record(final ReadableMap options, final Promise promise) {
+        if (mRecordingPromise == null) {
+            mCamera = RCTCamera.getInstance().acquireCameraInstance(options.getInt("type"));
+            if (null == mCamera) {
+                promise.reject("No camera found.");
+                return;
+            }
+            if (!prepareMediaRecorder(options)) {
+                promise.reject("Fail in prepareMediaRecorder()!");
```
Comment (markomatic) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java]: All promise rejections should include Throwable object as argument.

**Comment by jiw0220** [comment]
Exception in thread

Android:

Nexus 5x

"dependencies": {
    "react": "15.2.1",
    "react-native": "0.29.0",
    "react-native-camera": "git+https://github.com/marcejohnson/react-native-camera.git"
}

An error occurred while setting the "Camera.constants.CaptureQuality.low"

08-11 17:37:26.151 27524-27524/com.helloworld E/MediaRecorder: start failed: -2147483648
08-11 17:37:26.167 27524-16800/com.helloworld E/ReactNativeJS: { [Error: Exception in thread] framesToPop: 1, code: 'EUNSPECIFIED' }

**Comment by markomatic** [comment]
I will investigate this. I see now that for me it throws on picture capturing, but recording is fine when quality is low.

File: android/src/main/java/com/lwansbrough/RCTCamera/RCTCamera.java [hunk 10]
```diff
@@ -131,6 +163,46 @@ public int getPreviewHeight(int type) {
...
+                    if (newArea < resultArea) {
+                        result = size;
+                    }
+                }
+            }
+        }
+        return result;
+    }
+
+    public List<Camera.Size> getSupportedSizesWithFallback(Camera camera)
```
Comment (markomatic) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCamera.java]: This method is used for picture capturing also, but getSupportedPistureSizes is not used.

**Comment by markomatic** [comment]
@jiw0220 I don't know if your issue is fixed. If you have time, you can debug module. It is known issue on some devices (https://www.google.hr/search?q="start+failed%3A+-2147483648") and reason is unknown. I can't fix issue as it's not present on any of my devices.

**Comment by marcejohnson** [comment]
FYI merged commits from @markomatic (thanks Marko!) to address issues discussed above.  Please pull in latest and test to see if your particular issue(s) resolved.  Android camera is challenging cross-platform so appreciate the testing and fixes across devices.

**Comment by corbt** [comment]
FYI I'm having the same issue described above with the new commits, will continue investigating today.

**Comment by corbt** [comment]
@markomatic yes I agree, it appears the recording wasn't properly started in my case. No logs are generated on the JS or ADB sides at recording start though. I'll implement a [MediaRecorder.OnErrorListener](https://developer.android.com/reference/android/media/MediaRecorder.OnErrorListener.html) though which I suspect will more accurately pinpoint the problem.

File: android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java [hunk 11]
```diff
@@ -175,6 +197,211 @@ public String getName() {
...
+        }
+
+        mMediaRecorder.setOrientationHint(mediaRecorderHintOrientation);
+
+        if (cm == null) {
+            return new RuntimeException("CamcorderProfile not found in prepareMediaRecorder.");
+        }
+
+        cm.fileFormat = MediaRecorder.OutputFormat.MPEG_4;
+        cm.videoCodec = MediaRecorder.VideoEncoder.MPEG_4_SP;
```
Comment (corbt) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java]: Ok, I was able to fix my no-recording problem by commenting out this line (fix tested on Nexus 9 and Oneplus One). It appears my hardware doesn't have support for the `MPEG_4_SP` codec. Is this line absolutely necessary on some configurations, or should we just leave it out and let the system choose the default codec?

Comment (gmaclennan) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java]: Some details on options here: https://developer.android.com/guide/appendix/media-formats.html
(tl;dr MPEG_4_SP encoding is not supported by default in Android, it's a device specific implementation)

Comment (markomatic) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java]: Then this should be removed. That could be the same reason for @jiw0220 issue.

**Comment by marcejohnson** [comment]
Tonite merged commits from @corbt (thanks!) to address two issues discussed above (don't force specific video codec, save to correct cache dir).  Please pull in latest and test to see if your particular issue(s) are resolved.

**Comment by tannerkrewson** [comment]
I tested the latest on my Nexus 6 on 6.0 just like before. Still works great, however the video is still choppy.

**Comment by marcejohnson** [comment]
@tannerkrewson thanks for testing.  By default capture quality is high.  Did setCaptureVideoQuality at "medium" make replay less choppy?

File: android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java [hunk 12]
```diff
@@ -342,4 +610,15 @@ private File getTempMediaFile(int type) {
...
+
+    private void addToMediaStore(Uri uri) {
+        if (uri == null) {
+            Log.e(TAG, "Tried to store null url to media store.");
+            return;
+        }
+
+        Intent mediaScannerIntent = new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE);
+        mediaScannerIntent.setData(uri);
+        _reactContext.sendBroadcast(mediaScannerIntent);
```
Comment (AndrewJack) [inline android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java]: Should use MediaScannerConnection instead.

https://developer.android.com/reference/android/media/MediaScannerConnection.html

**Comment by jiw0220** [comment]
To solve the two problems In the nexus 5 x


1.Video the wrong direction

>   int mediaRecorderHintOrientation = (90 + ((720 - RCTCamera.getInstance().getActualDeviceOrientation() * 90))) % 360;
        // adjust for differences in how devices display front facing http://www.theverge.com/2015/11/9/9696774/google-nexus-5x-upside-down-camera
        if (RCT_CAMERA_TYPE_FRONT == options.getInt("type")) {
            if ((RCT_CAMERA_ORIENTATION_PORTRAIT == RCTCamera.getInstance().getOrientation()) && (RCTCamera.getInstance().getAdjustedDeviceOrientation() == 90)) {
                mediaRecorderHintOrientation += 180;
            }
        }
        //mMediaRecorder.setOrientationHint(mediaRecorderHintOrientation);
**mMediaRecorder.setOrientationHint(RCTCamera.getInstance().getAdjustedDeviceOrientation());**


2.recoder start fail at "low,medium"

>    switch (captureQuality) {
            case RCTCameraModule.RCT_CAMERA_CAPTURE_QUALITY_LOW:
                videoSize = getSmallestSize(getSupportedVideoSizes(camera));
                **cm = CamcorderProfile.get(_cameraTypeToIndex.get(cameraType),** CamcorderProfile.QUALITY_480P);
                break;
            case RCTCameraModule.RCT_CAMERA_CAPTURE_QUALITY_MEDIUM:
                List<Camera.Size> sizes = getSupportedVideoSizes(camera);
                videoSize = sizes.get(sizes.size() / 2);
                **cm = CamcorderProfile.get(_cameraTypeToIndex.get(cameraType),** CamcorderProfile.QUALITY_720P);
                break;
            case RCTCameraModule.RCT_CAMERA_CAPTURE_QUALITY_HIGH:
                videoSize = getBestSize(getSupportedVideoSizes(camera), Integer.MAX_VALUE, Integer.MAX_VALUE);
        }

**Comment by marcejohnson** [comment]
Today merged code from @markomatic, @jiw0220 and @AndrewJack.  Now using MediaScannerConnection, and addressed wrong orientation and failing low/medium profiles for better compatibility with nexus 5, etc.  Wow thanks for these great improvements.  Please pull in latest and test to see if your respective issue(s) are resolved.

**Comment by jiw0220** [comment]
hi.

android add checkDeviceAuthorizationStatus at RCTCameraModule.java

>  private static Promise promiseForPermissions;

> //TODO cover android 6.0 permission
    private boolean hasPermissionsGranted(String[] permissions) {
        for (String permission : permissions) {
            if (ActivityCompat.checkSelfPermission(getCurrentActivity(), permission) != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }
        return true;
    }

    public static void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (requestCode == REQUEST_VIDEO_PERMISSIONS) {
            Promise promise = promiseForPermissions;
            for (int i = 0; i < permissions.length; i++) {
                if (grantResults[i] != PackageManager.PERMISSION_GRANTED) {
                    promise.resolve(false);//TODO no grant
                    return;
                }
            }
            promise.resolve(true);
        }
    }

    @ReactMethod
    public void checkDeviceAuthorizationStatus(final Promise promise) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (!hasPermissionsGranted(VIDEO_PERMISSIONS)) {
                promiseForPermissions = promise;
                getCurrentActivity().requestPermissions(VIDEO_PERMISSIONS, REQUEST_VIDEO_PERMISSIONS);
            } else {
                promise.resolve(true);
            }
        } else {
            promise.resolve(true);
        }
    }

**Comment by sscaff1** [comment]
Any updates?

**Comment by markomatic** [comment]
This PR could be merged as it is obviously tested, and new PR with checkDeviceAuthorizationStatus for Android should be opened with @jiw0220 proposal.

**Comment by marcejohnson** [comment]
I second @markomatic's motion to merge.  :-)

**Comment by charpeni** [comment]
I would like to say thank you to everyone for this amazing feature.

**Comment by LoseTheQuit** [comment]
Is it possible for this to work with react-native@0.30? i read a comment that said you could record video on android, is that possible? If so, this is the error I'm getting: undefined is not an object(evaluating 'NativeModule.CameraManager.Aspect') - I really would like to get this working. Is there a viable solution?

**Comment by marcejohnson** [comment]
@LoseTheQuit yes, good news -- Android support for recording video is now merged with lwansbrough/react-native-camera:master.  The issue you're describing is usually a setup or linking problem.  Verify you completed all the setup steps here:  https://github.com/lwansbrough/react-native-camera#android .  Also check for linking issues like discussed in this ticket: https://github.com/lwansbrough/react-native-camera/issues/220

**Comment by LoseTheQuit** [comment]
I've done all the steps:
https://github.com/LoseTheQuit/homebrew/blob/master/components/CoreCamera.js
@marcejohnson Can you take a look and see if I missed something. I followed all the instructions from the guide.

**Comment by markomatic** [comment]
@LoseTheQuit, have you done: "Add new RCTCameraPackage() to the list returned by the getPackages() method". I don't see that line in MainActivity?

**Comment by LoseTheQuit** [comment]
I got it working!

**Pull request merged**